reference-implementation/to-upstream-wpts/writable-streams/aborting.js

```diff
@@ -43,5 +43,193 @@ promise_test(t => {
+  writer.abort(passedReason);
+
+  assert_equals(recordedReason, passedReason);
+}, 'Aborting a WritableStream passes through the given reason');
+
+promise_test(t => {
+  const ws = new WritableStream();
+  const writer = ws.getWriter();
+
+  const passedReason = new Error('Sorry, it just wasn\'t meant to be.');
+  writer.abort(passedReason);
+
+  return Promise.all([
+    promise_rejects(t, new TypeError(), writer.write(), 'writing should reject with the given reason'),
+    promise_rejects(t, new TypeError(), writer.close(), 'closing should reject with the given reason'),
+    promise_rejects(t, new TypeError(), writer.abort(), 'aborting should reject with the given reason'),
+    promise_rejects(t, new TypeError(), writer.closed, 'closed should reject with the given reason'),
+  ]);
+}, 'Aborting a WritableStream puts it in an errored state, with stored error equal to the abort reason');
+
+// TODO return promise
+promise_test(t => {
+  const ws = new WritableStream();
+  const writer = ws.getWriter();
+
+  promise_rejects(t, new TypeError(), writer.write('a'), 'writing should reject with a TypeError');
+
+  const passedReason = new Error('Sorry, it just wasn\'t meant to be.');
+  writer.abort(passedReason);
+}, 'Aborting a WritableStream causes any outstanding write() promises to be rejected with the abort reason');
+
+promise_test(t => {
+  const ws = new WritableStream();
+  const writer = ws.getWriter();
+
+  writer.close();
+
+  const passedReason = new Error('Sorry, it just wasn\'t meant to be.');
+  writer.abort(passedReason);
+
+  return promise_rejects(t, new TypeError('Aborted'), writer.closed, 'the stream should be errored with a TypeError');
+}, 'Closing but then immediately aborting a WritableStream causes the stream to error');
+
+promise_test(t => {
+  const ws = new WritableStream({
+    close() {
+      return new Promise(() => { }); // forever-pending
+    }
+  });
+
+  const writer = ws.getWriter();
+
+  writer.close();
+
+  const passedReason = new Error('Sorry, it just wasn\'t meant to be.');
+
+  setTimeout(() => {
+    writer.abort(passedReason);
+  }, 20);
+
+  return promise_rejects(t, new TypeError(), writer.closed, 'the stream should be errored with a TypeError');
+}, 'Closing a WritableStream and aborting it while it closes causes the stream to error');
+
+promise_test(() => {
+  const ws = new WritableStream();
+  const writer = ws.getWriter();
+
+  writer.close();
+
+  return delay(0)
+    .then(() => writer.abort())
+    .then(
+      v => assert_equals(v, undefined, 'abort promise should fulfill with undefined'),
+      () => { throw new Error(); }
+    );
+}, 'Aborting a WritableStream after it is closed is a no-op');
+
+test(() => {
+  const ws = new WritableStream({
+    close(...args) {
+      assert_equals(args.length, 0, 'close() was called (with no arguments)');
+    }
+  });
+
+  const writer = ws.getWriter();
+
+  writer.abort();
```

Comment:
Again best to use recordingWritableStream. As-is, this test will pass even if close is never called.

It might be good to add something to recordingWritableStream's close() that asserts that `arguments.length === 0`.